File: backend/app/api/excursions.py
"""Excursions API (Step 8)."""

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy.orm import Session

from app.db.models.market_data import TradeExcursion
from app.db.session import get_db
from app.services.excursion_enrichment.coverage import boundary_spread_stats, get_excursion_coverage, storage_stats
from app.services.excursion_enrichment.service import ExcursionEnrichmentService
from app.market_data.registry import get_market_data_provider

router = APIRouter(prefix="/api/excursions", tags=["excursions"])


class EnrichRequest(BaseModel):
    scope: str = "missing"
    dry_run: bool = False


@router.get("/coverage")
def coverage(db: Session = Depends(get_db)):
    return {
        **get_excursion_coverage(db),
        **boundary_spread_stats(db),
        **storage_stats(db),
    }


@router.post("/enrich")
def enrich(body: EnrichRequest, db: Session = Depends(get_db)):
    svc = ExcursionEnrichmentService(db, get_market_data_provider())
    return svc.enrich(scope=body.scope, dry_run=body.dry_run)


@router.post("/recalculate")
def recalculate(db: Session = Depends(get_db)):
    svc = ExcursionEnrichmentService(db, get_market_data_provider())
    return svc.recalculate()


@router.get("/trades/{trade_id}")
def get_trade_excursion(trade_id: int, db: Session = Depends(get_db)):
    row = db.query(TradeExcursion).filter(TradeExcursion.trade_id == trade_id).first()
    if not row:
        raise HTTPException(status_code=404, detail="Excursion data not found")
    return _serialize(row)


def _serialize(r: TradeExcursion) -> dict:
    import json

    flags = json.loads(r.quality_flags_json) if r.quality_flags_json else []
    return {
        "trade_id": r.trade_id,
        "data_provider": r.data_provider,
        "data_feed": r.data_feed,
        "data_resolution": r.data_resolution,
        "is_consolidated": r.is_consolidated,
        "holding_start_utc": r.holding_start_utc.isoformat() if r.holding_start_utc else None,
        "holding_end_utc": r.holding_end_utc.isoformat() if r.holding_end_utc else None,
        "reference_entry_price": str(r.reference_entry_price) if r.reference_entry_price else None,
        "price_mfe": str(r.price_mfe) if r.price_mfe is not None else None,
        "price_mae": str(r.price_mae) if r.price_mae is not None else None,
        "price_mfe_pct": str(r.price_mfe_pct) if r.price_mfe_pct is not None else None,
        "price_mae_pct": str(r.price_mae_pct) if r.price_mae_pct is not None else None,
        "conservative_price_mfe": str(r.conservative_price_mfe) if r.conservative_price_mfe is not None else None,
        "conservative_price_mae": str(r.conservative_price_mae) if r.conservative_price_mae is not None else None,
        "position_mfe_amount": str(r.position_mfe_amount) if r.position_mfe_amount is not None else None,
        "position_mae_amount": str(r.position_mae_amount) if r.position_mae_amount is not None else None,
        "mfe_r": str(r.mfe_r) if r.mfe_r is not None else None,
        "mae_r": str(r.mae_r) if r.mae_r is not None else None,
        "conservative_position_mfe_amount": str(r.conservative_position_mfe_amount) if r.conservative_position_mfe_amount is not None else None,
        "conservative_position_mae_amount": str(r.conservative_position_mae_amount) if r.conservative_position_mae_amount is not None else None,
        "conservative_mfe_r": str(r.conservative_mfe_r) if r.conservative_mfe_r is not None else None,
        "conservative_mae_r": str(r.conservative_mae_r) if r.conservative_mae_r is not None else None,
        "mfe_boundary_spread_amount": str(r.mfe_boundary_spread_amount) if r.mfe_boundary_spread_amount is not None else None,
        "mfe_boundary_spread_r": str(r.mfe_boundary_spread_r) if r.mfe_boundary_spread_r is not None else None,
        "mae_boundary_spread_amount": str(r.mae_boundary_spread_amount) if r.mae_boundary_spread_amount is not None else None,
        "mae_boundary_spread_r": str(r.mae_boundary_spread_r) if r.mae_boundary_spread_r is not None else None,
        "mfe_time_utc": r.mfe_time_utc.isoformat() if r.mfe_time_utc else None,
        "mae_time_utc": r.mae_time_utc.isoformat() if r.mae_time_utc else None,
        "time_to_mfe_seconds": r.time_to_mfe_seconds,
        "time_to_mae_seconds": r.time_to_mae_seconds,
        "mfe_to_exit_seconds": r.mfe_to_exit_seconds,
        "gross_realized_pnl": str(r.gross_realized_pnl) if r.gross_realized_pnl is not None else None,
        "gross_realized_r": str(r.gross_realized_r) if r.gross_realized_r is not None else None,
        "exit_efficiency_pct": str(r.exit_efficiency_pct) if r.exit_efficiency_pct is not None else None,
        "r_left_on_table": str(r.r_left_on_table) if r.r_left_on_table is not None else None,
        "peak_giveback_amount": str(r.peak_giveback_amount) if r.peak_giveback_amount is not None else None,
        "peak_giveback_r": str(r.peak_giveback_r) if r.peak_giveback_r is not None else None,
        "peak_giveback_pct": str(r.peak_giveback_pct) if r.peak_giveback_pct is not None else None,
        "post_exit_favorable_5m": str(r.post_exit_favorable_5m) if r.post_exit_favorable_5m is not None else None,
        "post_exit_favorable_15m": str(r.post_exit_favorable_15m) if r.post_exit_favorable_15m is not None else None,
        "post_exit_favorable_30m": str(r.post_exit_favorable_30m) if r.post_exit_favorable_30m is not None else None,
        "post_exit_favorable_5m_r": str(r.post_exit_favorable_5m_r) if r.post_exit_favorable_5m_r is not None else None,
        "post_exit_favorable_15m_r": str(r.post_exit_favorable_15m_r) if r.post_exit_favorable_15m_r is not None else None,
        "post_exit_favorable_30m_r": str(r.post_exit_favorable_30m_r) if r.post_exit_favorable_30m_r is not None else None,
        "copilot_exit_time_utc": r.copilot_exit_time_utc.isoformat() if r.copilot_exit_time_utc else None,
        "copilot_exit_price": str(r.copilot_exit_price) if r.copilot_exit_price is not None else None,
        "copilot_exit_delta_seconds": r.copilot_exit_delta_seconds,
        "copilot_exit_delta_price": str(r.copilot_exit_delta_price) if r.copilot_exit_delta_price is not None else None,
        "copilot_exit_delta_pct": str(r.copilot_exit_delta_pct) if r.copilot_exit_delta_pct is not None else None,
        "quality_status": r.quality_status,
        "quality_flags": flags,
        "boundary_ambiguity": r.boundary_ambiguity,
        "efficiency_over_100": r.efficiency_over_100,
        "sparse_interval": r.sparse_interval,
        "longest_bar_gap_seconds": r.longest_bar_gap_seconds,
        "provider_missing_data": r.provider_missing_data,
        "calculation_version": r.calculation_version,
        "calculated_at": r.calculated_at.isoformat() if r.calculated_at else None,
    }

File: frontend/src/api/excursions.ts
import type { DashboardFiltersState } from '../types/dashboard';
import { filtersToQueryParams } from '../utils/dates';

import { apiRequest as request } from './base';

export interface ExcursionCoverage {
  total_closed_trades: number;
  excursion_enriched: number;
  excursion_coverage_pct: number | null;
  r_qualified_excursions: number;
  mfe_r_coverage_pct: number | null;
  consolidated_count: number;
  partial_feed_count: number;
  boundary_ambiguous_count: number;
  sparse_interval_count: number;
  missing_count: number;
  copilot_exit_available: number;
  copilot_coverage_pct: number | null;
  copilot_status: string;
  intraday_bars_cached: number;
  unique_symbol_days: number;
  avg_mfe_spread_amount?: number | null;
  avg_mfe_spread_r?: number | null;
  median_mfe_spread_r?: number | null;
  p95_mfe_spread_r?: number | null;
  count_spread_gt_025r?: number;
  count_spread_gt_050r?: number;
  database_size_bytes?: number;
  database_size_mb?: number;
  intraday_bar_count?: number;
  avg_bars_per_symbol_day?: number;
  storage_advisory?: string | null;
}

export interface TradeExcursion {
  trade_id: number;
  data_provider: string | null;
  data_feed: string | null;
  data_resolution: string | null;
  is_consolidated: boolean | null;
  holding_start_utc: string | null;
  holding_end_utc: string | null;
  reference_entry_price: string | null;
  price_mfe: string | null;
  price_mae: string | null;
  conservative_price_mfe: string | null;
  conservative_price_mae: string | null;
  position_mfe_amount: string | null;
  position_mae_amount: string | null;
  mfe_r: string | null;
  mae_r: string | null;
  conservative_position_mfe_amount: string | null;
  conservative_position_mae_amount: string | null;
  conservative_mfe_r: string | null;
  conservative_mae_r: string | null;
  mfe_boundary_spread_amount: string | null;
  mfe_boundary_spread_r: string | null;
  mae_boundary_spread_amount: string | null;
  mae_boundary_spread_r: string | null;
  mfe_time_utc: string | null;
  mae_time_utc: string | null;
  time_to_mfe_seconds: number | null;
  time_to_mae_seconds: number | null;
  mfe_to_exit_seconds: number | null;
  gross_realized_pnl: string | null;
  gross_realized_r: string | null;
  exit_efficiency_pct: string | null;
  r_left_on_table: string | null;
  peak_giveback_amount: string | null;
  peak_giveback_r: string | null;
  peak_giveback_pct: string | null;
  post_exit_favorable_5m: string | null;
  post_exit_favorable_15m: string | null;
  post_exit_favorable_30m: string | null;
  post_exit_favorable_5m_r: string | null;
  post_exit_favorable_15m_r: string | null;
  post_exit_favorable_30m_r: string | null;
  copilot_exit_time_utc: string | null;
  copilot_exit_price: string | null;
  copilot_exit_delta_seconds: number | null;
  copilot_exit_delta_price: string | null;
  copilot_exit_delta_pct: string | null;
  quality_status: string;
  quality_flags: string[];
  boundary_ambiguity: boolean;
  efficiency_over_100: boolean;
  sparse_interval: boolean;
  longest_bar_gap_seconds: number | null;
  provider_missing_data: boolean;
  calculation_version: string | null;
  calculated_at: string | null;
}

export interface ExitAnalysisSummary {
  average_mfe_r?: string | null;
  average_mae_r?: string | null;
  average_exit_efficiency?: string | null;
  median_exit_efficiency?: string | null;
  average_r_left_on_table?: string | null;
  average_peak_giveback_pct?: string | null;
  median_time_to_mfe_seconds?: string | null;
  capture_ge_25_pct?: number | null;
  capture_ge_50_pct?: number | null;
  capture_ge_75_pct?: number | null;
  capture_ge_90_pct?: number | null;
  positive_mfe_to_loss_count?: number;
  positive_mfe_to_loss_pct?: number | null;
  reached_2r_closed_lt_1r?: number;
  reached_2r_closed_losing?: number;
  best_capture_min_mfe_r?: string;
}

export interface ExitAnalysisScatterPoint {
  trade_id: number;
  mfe_r: string;
  mae_r: string | null;
  actual_r: string;
  ticker: string;
}

export interface ExitAnalysisTableRow {
  trade_id: number;
  ticker: string;
  exit_date: string | null;
  actual_r: string | null;
  mfe_r: string | null;
  mae_r: string | null;
  r_left_on_table: string | null;
  exit_efficiency_pct: string | null;
  peak_giveback_r: string | null;
  peak_giveback_pct: string | null;
  mfe_to_exit_seconds: number | null;
  quality_status: string;
}

export interface ExitAnalysisResponse {
  summary: ExitAnalysisSummary;
  coverage: ExcursionCoverage;
  scatter: ExitAnalysisScatterPoint[];
  worst_left_on_table: ExitAnalysisTableRow[];
  worst_giveback: ExitAnalysisTableRow[];
  best_capture: ExitAnalysisTableRow[];
}

export interface EnrichResult {
  status?: string;
  trades_requested?: number;
  enriched?: number;
  skipped?: number;
  errors?: number;
}

export function fetchExcursionCoverage(): Promise<ExcursionCoverage> {
  return request<ExcursionCoverage>('/excursions/coverage');
}

export function enrichExcursions(scope = 'missing', dryRun = false): Promise<EnrichResult> {
  return request<EnrichResult>('/excursions/enrich', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ scope, dry_run: dryRun }),
  });
}

export function recalculateExcursions(): Promise<EnrichResult> {
  return request<EnrichResult>('/excursions/recalculate', { method: 'POST' });
}

export function fetchTradeExcursion(tradeId: number): Promise<TradeExcursion> {
  return request<TradeExcursion>(`/excursions/trades/${tradeId}`);
}

export function fetchExitAnalysis(filters: DashboardFiltersState): Promise<ExitAnalysisResponse> {
  const params = new URLSearchParams();
  const qp = filtersToQueryParams(filters);
  Object.entries(qp).forEach(([k, v]) => {
    if (k !== 'range' && v) params.set(k, v);
  });
  return request<ExitAnalysisResponse>(`/exit-analysis?${params}`);
}
